torchdata/datapipes/iter/transform/callable.py

```diff
@@ -585,3 +591,200 @@ def __new__(
+
+            - ``None`` as default to apply ``fn`` to the data directly.
+            - Integer(s) is used for list/tuple.
+            - Key(s) is used for dict.
+
+        output_col: Index of data where result of ``fn`` is placed. ``output_col`` can be specified
+            only when ``input_col`` is not ``None``
+
+            - ``None`` as default to replace the index that ``input_col`` specified; For ``input_col`` with
+              multiple indices, the left-most one is used, and other indices will be removed.
+            - Integer is used for list/tuple. ``-1`` represents to append result at the end.
+            - Key is used for dict. New key is acceptable.
+
+        scheduled_tasks: How many tasks will be scheduled at any given time (Default value: 64)
+        max_workers: Maximum number of threads to execute function calls
+        **threadpool_kwargs: additional arguments to be given to the ``ThreadPoolExecutor``
+
+    Note:
+         For more information about ``max_workers`` and additional arguments for the ``ThreadPoolExecutor``
+         please refer to: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor
+
+    Note:
+        For optimal use of all threads, we recommend ``scheduled_tasks`` > ``max_workers``. High value of ``scheduled_tasks``
+        might lead to long waiting period until the first element is yielded as ``next`` is called
+        ``scheduled_tasks`` many times on ``source_datapipe``  before yielding.
+
+    Example:
+
+    .. testsetup::
+
+        from torchdata.datapipes.iter import IterableWrapper
+        import requests
+        import time
+        from unittest.mock import MagicMock
+
+        requests.get = MagicMock()
+        urls = []
+
+    .. testcode::
+
+        # fetching html from remote
+        def fetch_html(url: str, **kwargs):
+            r = requests.get(url, **kwargs)
+            r.raise_for_status()
+            return r.content
+        dp = IterableWrapper(urls)
+        dp = dp.threadpool_map(fetch_html,max_workers=16)
+
+    .. testcode::
+
+        def mul_ten(x):
+            time.sleep(0.1)
+            return x * 10
+
+        dp = IterableWrapper([(i, i) for i in range(50)])
+        dp = dp.threadpool_map(mul_ten, input_col=1)
+        print(list(dp))
+
+    .. testoutput::
+
+        [(0, 0), (1, 10), (2, 20), (3, 30), ...]
+
+    .. testcode::
+
+        dp = IterableWrapper([(i, i) for i in range(50)])
+        dp = dp.threadpool_map(mul_ten, input_col=1, output_col=-1)
+        print(list(dp))
+
+    .. testoutput::
+
+        [(0, 0, 0), (1, 1, 10), (2, 2, 20), (3, 3, 30), ...]
+
+    """
+
+    datapipe: IterDataPipe
+    fn: Callable
+
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        fn: Callable,
+        input_col=None,
+        output_col=None,
+        scheduled_tasks: int = 64,
+        max_workers: Optional[int] = None,
```

Comment:
I think now we can consolidate those two arguments to a single one. To align with `ThreadPoolExecutor`, let's keep `max_workers` then.

Reply:
How exactly? If we want to leave `max_workers` as `None` by default (as is the case in ThreadPoolExecutor, "Changed in version 3.8: Default value of max_workers is changed to min(32, os.cpu_count() + 4")
how would we choose `scheduled_tasks`? Set it to `2*threads` if `threads` is specified and some default value otherwise?

Reply:
I am thinking with a default number of `max_workers`, we can always schedule the same amount of tasks to maximize the number of workers, right?

Reply:
It should be higher as `max_workers` as not all tasks will take the same amount of time. With `max_workers` = `scheduled_tasks` some threads will sit idle. I think it depends on high big the variance of needed time for the given function is. As this is something that only the user knows we can either force them to choose or set a high enough default value.

Reply:
If that's the case, should we turn default value of `scheduled_tasks` to 128?

Reply:
Setting a higher default value for `scheduled_tasks` is probably a good idea.  As the impact of both `workers` and `scheduled_tasks` probably (heavily) depends on the function and the software/hardware that is used I would refrain from making too concrete suggestions. 
Instead I think adding a more detailed comment about the impact of `scheduled_tasks` and encouraging users to try out different values for optimal performance should be sufficient.

Reply:
I have updated the default value and added a remark about the impact of `scheduled_tasks`. We already refer to https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor in the note before. Let me know if that is still insufficient.

Reply:
I am going to land this PR shortly. I just took a deeper look at the doc of `ThreadPoolExecutor`, it seems the default `ma_workers` is `min(32, os.cpu_count() + 4)`. Giving 128 scheduled tasks might not be ideal. But, at least it's not a blocking issue, I would like to land it and fix the `scheduled_tasks` when users give us more specific performance feedback on it.

Reply:
The default value is stilll up to change, yes. It's only necessary as the normal ThreadPool `map` would first iterate over the whole dp to submit all tasks. To avoid this we use `scheduled_tasks`, so it only affects how many tasks will be put in a queue (submitted).

The threads itself do not switch context until they finished their current task. Therefore the number of threads and `scheduled_tasks` only interact with each other if the variance of the time needed to complete the task is high. The main thread waits on the result of the thread that is working on the next (in correct order) result. Meanwhile the other threads can already work on other already submitted tasks. At least that's what I make of the source code of ThreadPoolExecutor, see: https://github.com/python/cpython/blob/1f7eb8fe6c1a566caf934a96c6fe2877b2447d5b/Lib/concurrent/futures/thread.py#L69

Reply:
I see what do you mean. The tasks are not executed until any worker become available.

Another point I want to make is the higher `scheduled_tasks`, the more memory is needed to load data from prior `DataPipe`. This is not a problem for the common use cases, as this DataPipe should be normally used to perform I/O operations, where input is just paths/urls.